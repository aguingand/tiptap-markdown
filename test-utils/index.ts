
export { parse } from './parse.js';
export { serialize } from './serialize';

export function dedent(str: TemplateStringsArray) {
    return str[0]!.replace(/^\s*/gm, '');
}
