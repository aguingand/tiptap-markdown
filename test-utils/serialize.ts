import { Node as ProseMirrorNode } from "@tiptap/pm/model";
import { Content, getHTMLFromFragment, createNodeFromContent } from "@tiptap/core";
import { createEditor, TestEditorOptions } from "./editor";

export function serialize(content: Content, options: TestEditorOptions = {}) {
    const editor = createEditor(options);
    const doc = createNodeFromContent(content, editor.schema, {
        parseOptions: {
            preserveWhitespace: true,
        }
    }) as ProseMirrorNode;
    const html = getHTMLFromFragment(doc.content, editor.schema);

    return editor.storage.markdown.serializer.serialize(html);
}
