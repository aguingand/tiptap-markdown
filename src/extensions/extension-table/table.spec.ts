import { describe, expect, test } from "vitest";
import { dedent, parse, serialize } from "../../../test-utils";


test('parse markdown', () => {
    expect(parse(dedent`
        example1 | example2
        --- | ---
        example3 | example4
    `)).toMatchSnapshot();
});

test('parse html', () => {
    expect(parse(dedent`
        <table>
        <thead>
            <tr>
                <th>example1</th>
                <th>example2</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>example3</td>
                <td>example4</td>
            </tr>
        </tbody>
        </table>
    `)).toMatchSnapshot();
});

describe('serialize', () => {
    test('filled', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th>example1</th>
                    <th>example2</th>
                </tr>
                <tr>
                    <td>example3</td>
                    <td>example4</td>
                </tr>
            </table>
        `)).toEqual(dedent`
            | example1 | example2 |
            | -------- | -------- |
            | example3 | example4 |
        `);
    });
    test('empty', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th></th>
                    <th></th>
                </tr>
                <tr>
                    <td></td>
                    <td></td>
                </tr>
            </table>
        `)).toEqual(dedent`
            |   |   |
            | - | - |
            |   |   |
        `);
    });
    test('single column', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th>example1</th>
                </tr>
                <tr>
                    <td>example3</td>
                </tr>
            </table>
        `)).toEqual(dedent`
            | example1 |
            | -------- |
            | example3 |
        `);
    });
    test('header only', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th>example1</th>
                    <th>example2</th>
                </tr>
            </table>
        `)).toEqual(dedent`
            | example1 | example2 |
            | -------- | -------- |
        `);
    });
    test('cell with hard break', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th><p>example1 <br> example2</p></th>
                </tr>
            </table>
        `, { markdown: { html: true } })).toEqual(dedent`
            | example1<br>example2 |
            | -------------------- |
        `);
    });
    test('no header', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <td>example1</td>
                </tr>
                <tr>
                    <td>example3</td>
                </tr>
            </table>
        `, { markdown: { html: true } })).toMatchSnapshot();
    });
    test('header in body', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th>example1</th>
                </tr>
                <tr>
                    <th>example3</th>
                </tr>
            </table>
        `, { markdown: { html: true } })).toMatchSnapshot();
    });
    test('with colspan', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th colspan="2">example1</th>
                </tr>
            </table>
        `, { markdown: { html: true } })).toMatchSnapshot();
    });
    test('with rowspan', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th rowspan="2">example1</th>
                </tr>
            </table>
        `, { markdown: { html: true } })).toMatchSnapshot();
    });
    test('multiline cell', () => {
        expect(serialize(dedent`
            <table>
                <tr>
                    <th><p>example1</p><p>example2</p></th>
                </tr>
            </table>
        `, { markdown: { html: true } })).toMatchSnapshot();
    });
})
