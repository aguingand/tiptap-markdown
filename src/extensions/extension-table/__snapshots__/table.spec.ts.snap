// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse html 1`] = `
[
  {
    "content": [
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example1",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example2",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
        ],
        "type": "tableRow",
      },
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example3",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example4",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
        ],
        "type": "tableRow",
      },
    ],
    "type": "table",
  },
]
`;

exports[`parse markdown 1`] = `
[
  {
    "content": [
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example1",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example2",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
        ],
        "type": "tableRow",
      },
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example3",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example4",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
        ],
        "type": "tableRow",
      },
    ],
    "type": "table",
  },
]
`;
